### PR TITLE
Replace a long inline "autoref" comment with method docs

### PR DESCRIPTION
**Pull request by Zalathar**
This comment has two problems:

- It is very long, making the flow of the enclosing method hard to follow.
- It starts by talking about an `autoref` flag that hasn't existed since #59114.
  - This makes it hard to trust that the information in the comment is accurate or relevant, even though much of it still seems to be true.

This PR therefore replaces the long inline comment with a revised doc comment on `bind_matched_candidate_for_guard`, and some shorter inline comments.

For readers who want more historical context, we also link to the PR that added the old comment, and the PR that removed the `autoref` flag.

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/matches/mod.rs [hunk 1]
```diff
@@ -2177,87 +2177,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
-        // match place { ref mut foo if inspect(foo)
-        //     => feed(foo), ... }
-        // ```
-        //
-        // will be treated as if it were really something like:
-        //
-        // ```
-        // let place = Foo::new();
-        // match place { Foo { .. } if { let tmp1 = & &mut place; inspect(*tmp1) }
-        //     => { let tmp2 = &mut place; feed(tmp2) }, ... }
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:2222]: Could you keep these examples around? The `&&mut` business isn't obvious, I found the examples helpful

Comment (Zalathar) [reply compiler/rustc_mir_build/src/build/matches/mod.rs:2222]: Yeah, I've now transplanted these examples into the docs for `bind_matched_candidate_for_guard`, so they're still visible but don't get in the way of reading any method body.

Hmm, maybe I should move over the “implementation notes” section as well (lightly edited).

**Comment by Nadrieril** [comment]
Looks good!

@bors r+ rollup
